addons/font/bmfont.c

```diff
@@ -0,0 +1,498 @@
+#include "allegro5/allegro.h"
+#include "allegro5/allegro_font.h"
+#include <stdio.h>
+
+#include "allegro5/internal/aintern_font.h"
+
+#include "font.h"
+#include "xml.h"
+
+ALLEGRO_DEBUG_CHANNEL("font")
+
+typedef struct {
+   int first;
+   int second;
+   int amount;
+} BMFONT_KERNING;
+
+typedef struct {
+   int page;
+   int x, y;
+   int width, height;
+   int xoffset, yoffset;
+   int xadvance;
+   int chnl;
+   int kerning_pairs;
+   BMFONT_KERNING *kerning;
+} BMFONT_CHAR;
+
+typedef struct BMFONT_RANGE BMFONT_RANGE;
+
+struct BMFONT_RANGE {
+   int first;
+   int count;
+   BMFONT_CHAR **characters;
+   BMFONT_RANGE *next;
+};
+
+typedef struct {
+   ALLEGRO_USTR *tag;
+   ALLEGRO_USTR *attribute;
+   int pages_count;
+   ALLEGRO_BITMAP **pages;
+   BMFONT_RANGE *range_first;
```

Comment:
Curious, why use a linked list?

Reply:
Isn't that what Allegro's normal bitmap font uses? Thinking about it it might be possible to use a binary tree or something for faster lookup. My use case did not have a lot of separate ranges so a linked list was good enough.